File: server/src/routes/communication.rs
use axum::{
    extract::{
        ws::{Message, WebSocket, WebSocketUpgrade},
        Path, Query, State,
    },
    http::HeaderMap,
    response::IntoResponse,
    routing::get,
    Json, Router,
};
use serde::Deserialize;
use uuid::Uuid;

use crate::error::AppError;
use crate::http_auth::auth_user;
use crate::models::communication::{
    MailboxListResponse, PatchMailboxRequest, SendMessageRequest, SendMessageResponse,
    UnreadCountResponse,
};
use crate::repos::communication;
use crate::state::AppState;

#[derive(Deserialize)]
pub struct ListQuery {
    folder: String,
    #[serde(default)]
    q: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct WsAuthMessage {
    auth_token: String,
}

fn notify_mailbox(state: &AppState, user_id: Uuid) {
    let _ = state
        .comm_events
        .send((user_id, r#"{"type":"mailbox_updated"}"#.to_string()));
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route(
            "/api/v1/communication/messages",
            get(list_handler).post(send_handler),
        )
        .route(
            "/api/v1/communication/messages/{id}",
            get(get_handler).patch(patch_handler),
        )
        .route("/api/v1/communication/unread-count", get(unread_handler))
        .route("/api/v1/communication/ws", get(ws_handler))
}

fn validate_folder(folder: &str) -> Result<(), AppError> {
    match folder {
        "inbox" | "starred" | "sent" | "drafts" | "trash" => Ok(()),
        _ => Err(AppError::InvalidInput("Invalid folder.".into())),
    }
}

fn validate_mailbox_folder(folder: &str) -> Result<(), AppError> {
    match folder {
        "inbox" | "sent" | "drafts" | "trash" => Ok(()),
        _ => Err(AppError::InvalidInput("Invalid folder.".into())),
    }
}

async fn list_handler(
    State(state): State<AppState>,
    headers: HeaderMap,
    Query(q): Query<ListQuery>,
) -> Result<Json<MailboxListResponse>, AppError> {
    let user = auth_user(&state, &headers)?;
    validate_folder(&q.folder)?;
    let messages = communication::list_for_user(&state.pool, user.user_id, &q.folder, &q.q).await?;
    Ok(Json(MailboxListResponse { messages }))
}

async fn get_handler(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    headers: HeaderMap,
) -> Result<Json<crate::models::communication::MailboxMessage>, AppError> {
    let user = auth_user(&state, &headers)?;
    let msg = communication::get_for_user(&state.pool, user.user_id, id)
        .await?
        .ok_or(AppError::NotFound)?;
    Ok(Json(msg))
}

async fn unread_handler(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> Result<Json<UnreadCountResponse>, AppError> {
    let user = auth_user(&state, &headers)?;
    let unread_inbox = communication::count_unread_inbox(&state.pool, user.user_id).await?;
    Ok(Json(UnreadCountResponse { unread_inbox }))
}

async fn send_handler(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(req): Json<SendMessageRequest>,
) -> Result<Json<SendMessageResponse>, AppError> {
    let user = auth_user(&state, &headers)?;
    let subject = req.subject.trim();
    let body = req.body.trim();

    if req.draft {
        let id = communication::save_draft(&state.pool, user.user_id, subject, body).await?;
        notify_mailbox(&state, user.user_id);
        return Ok(Json(SendMessageResponse { id }));
    }

    let to_email = req
        .to_email
        .as_ref()
        .map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .ok_or_else(|| AppError::InvalidInput("to_email is required to send.".into()))?;

    let id = communication::send_message(&state.pool, user.user_id, to_email, subject, body)
        .await?
        .ok_or_else(|| AppError::InvalidInput("No user registered with that email.".into()))?;

    notify_mailbox(&state, user.user_id);
    if let Some(recipient) = crate::repos::user::find_by_email(&state.pool, to_email).await? {
        if recipient.id != user.user_id {
            notify_mailbox(&state, recipient.id);
        }
    }

    Ok(Json(SendMessageResponse { id }))
}

async fn patch_handler(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    headers: HeaderMap,
    Json(req): Json<PatchMailboxRequest>,
) -> Result<Json<serde_json::Value>, AppError> {
    let user = auth_user(&state, &headers)?;
    if let Some(ref f) = req.folder {
        validate_mailbox_folder(f)?;
    }
    let ok = communication::update_mailbox(&state.pool, user.user_id, id, &req).await?;
    if !ok {
        return Err(AppError::NotFound);
    }
    notify_mailbox(&state, user.user_id);
    Ok(Json(serde_json::json!({ "ok": true })))
}

/// `WebSocketUpgrade` must be the last extractor or the handshake will not complete.
async fn ws_handler(
    State(state): State<AppState>,
    ws: WebSocketUpgrade,
) -> Result<impl IntoResponse, AppError> {
    Ok(ws.on_upgrade(move |socket| handle_ws(socket, state)))
}

async fn read_ws_user_id(socket: &mut WebSocket, state: &AppState) -> Option<Uuid> {
    let text = loop {
        match socket.recv().await {
            Some(Ok(Message::Text(t))) => break t,
            Some(Ok(Message::Ping(p))) => {
                let _ = socket.send(Message::Pong(p)).await;
            }
            Some(Ok(Message::Close(_))) | None | Some(Err(_)) => return None,
            _ => {}
        }
    };
    let auth: WsAuthMessage = match serde_json::from_str(&text) {
        Ok(a) => a,
        Err(_) => return None,
    };
    state.jwt.verify(&auth.auth_token).ok().map(|u| u.user_id)
}

async fn handle_ws(mut socket: WebSocket, state: AppState) {
    let Some(user_id) = read_ws_user_id(&mut socket, &state).await else {
        let _ = socket.send(Message::Close(None)).await;
        return;
    };
    let mut rx = state.comm_events.subscribe();
    loop {
        tokio::select! {
            msg = socket.recv() => {
                match msg {
                    Some(Ok(Message::Close(_))) | None => break,
                    Some(Ok(Message::Ping(p))) => {
                        let _ = socket.send(Message::Pong(p)).await;
                    }
                    Some(Err(_)) => break,
                    _ => {}
                }
            }
            r = rx.recv() => {
                match r {
                    Ok((uid, text)) if uid == user_id => {
                        if socket.send(Message::Text(text.into())).await.is_err() {
                            break;
                        }
                    }
                    Err(tokio::sync::broadcast::error::RecvError::Lagged(_)) => {}
                    Err(tokio::sync::broadcast::error::RecvError::Closed) => break,
                    Ok(_) => {}
                }
            }
        }
    }
}

File: server/internal/httpserver/communication.go
package httpserver

import (
	"context"
	"encoding/json"
	"net/http"
	"strings"
	"time"

	"github.com/coder/websocket"
	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"
	"github.com/lextures/lextures/server/internal/apierr"
	commodels "github.com/lextures/lextures/server/internal/models/communication"
	"github.com/lextures/lextures/server/internal/repos/communication"
	"github.com/lextures/lextures/server/internal/repos/user"
)

const mailboxUpdateJSON = `{"type":"mailbox_updated"}`

func validateListFolder(f string) bool {
	switch f {
	case "inbox", "starred", "sent", "drafts", "trash":
		return true
	default:
		return false
	}
}

func validatePatchFolder(f string) bool {
	switch f {
	case "inbox", "sent", "drafts", "trash":
		return true
	default:
		return false
	}
}

func (d Deps) notifyMailbox(userID uuid.UUID) {
	if d.Comm == nil {
		return
	}
	d.Comm.Broadcast(userID, mailboxUpdateJSON)
}

// handleCommMessagesList is GET /api/v1/communication/messages?folder&...
func (d Deps) handleCommMessagesList() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet {
			w.Header().Set("Allow", http.MethodGet)
			http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
			return
		}
		userID, ok := d.meUserID(w, r)
		if !ok {
			return
		}
		folder := r.URL.Query().Get("folder")
		if folder == "" || !validateListFolder(folder) {
			apierr.WriteJSON(w, http.StatusBadRequest, apierr.CodeInvalidInput, "Invalid folder.")
			return
		}
		q := r.URL.Query().Get("q")
		msgs, err := communication.ListForUser(r.Context(), d.Pool, userID, folder, q)
		if err != nil {
			apierr.WriteJSON(w, http.StatusInternalServerError, apierr.CodeInternal, "Could not list messages.")
			return
		}
		if msgs == nil {
			msgs = []commodels.MailboxMessage{}
		}
		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(w).Encode(commodels.MailboxListResponse{Messages: msgs})
	}
}

// handleCommMessagesPost is POST /api/v1/communication/messages
func (d Deps) handleCommMessagesPost() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			w.Header().Set("Allow", http.MethodPost)
			http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
			return
		}
		userID, ok := d.meUserID(w, r)
		if !ok {
			return
		}
		var req struct {
			ToEmailSnake *string `json:"to_email"`
			ToEmailCamel *string `json:"toEmail"`
			Subject      string  `json:"subject"`
			Body         string  `json:"body"`
			Draft        bool    `json:"draft"`
		}
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			apierr.WriteJSON(w, http.StatusBadRequest, apierr.CodeInvalidInput, "Invalid JSON body.")
			return
		}
		req.Subject = strings.TrimSpace(req.Subject)
		req.Body = strings.TrimSpace(req.Body)
		if req.Draft {
			id, err := communication.SaveDraft(r.Context(), d.Pool, userID, req.Subject, req.Body)
			if err != nil {
				apierr.WriteJSON(w, http.StatusInternalServerError, apierr.CodeInternal, "Could not save draft.")
				return
			}
			d.notifyMailbox(userID)
			w.Header().Set("Content-Type", "application/json; charset=utf-8")
			_ = json.NewEncoder(w).Encode(commodels.SendMessageResponse{ID: id})
			return
		}
		to := ""
		if req.ToEmailSnake != nil {
			to = strings.TrimSpace(*req.ToEmailSnake)
		} else if req.ToEmailCamel != nil {
			to = strings.TrimSpace(*req.ToEmailCamel)
		}
		if to == "" {
			apierr.WriteJSON(w, http.StatusBadRequest, apierr.CodeInvalidInput, "to_email is required to send.")
			return
		}
		id, err := communication.SendMessage(r.Context(), d.Pool, userID, to, req.Subject, req.Body)
		if err != nil {
			apierr.WriteJSON(w, http.StatusInternalServerError, apierr.CodeInternal, "Could not send message.")
			return
		}
		if id == nil {
			apierr.WriteJSON(w, http.StatusBadRequest, apierr.CodeInvalidInput, "No user registered with that email.")
			return
		}
		d.notifyMailbox(userID)
		if u, e := user.FindByEmail(r.Context(), d.Pool, user.NormalizeEmail(to)); e == nil && u != nil {
			if rid, e2 := uuid.Parse(u.ID); e2 == nil && rid != userID {
				d.notifyMailbox(rid)
			}
		}
		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(w).Encode(commodels.SendMessageResponse{ID: *id})
	}
}

// handleCommMessageGet is GET /api/v1/communication/messages/{id}
func (d Deps) handleCommMessageGet() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet {
			w.Header().Set("Allow", http.MethodGet)
			http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
			return
		}
		userID, ok := d.meUserID(w, r)
		if !ok {
			return
		}
		id, err := uuid.Parse(chi.URLParam(r, "id"))
		if err != nil {
			apierr.WriteJSON(w, http.StatusBadRequest, apierr.CodeInvalidInput, "Invalid message id.")
			return
		}
		msg, err := communication.GetForUser(r.Context(), d.Pool, userID, id)
		if err != nil {
			apierr.WriteJSON(w, http.StatusInternalServerError, apierr.CodeInternal, "Could not load message.")
			return
		}
		if msg == nil {
			apierr.WriteJSON(w, http.StatusNotFound, apierr.CodeNotFound, "Not found.")
			return
		}
		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(w).Encode(msg)
	}
}

// handleCommMessagePatch is PATCH /api/v1/communication/messages/{id}
func (d Deps) handleCommMessagePatch() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPatch {
			w.Header().Set("Allow", http.MethodPatch)
			http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
			return
		}
		userID, ok := d.meUserID(w, r)
		if !ok {
			return
		}
		mid, err := uuid.Parse(chi.URLParam(r, "id"))
		if err != nil {
			apierr.WriteJSON(w, http.StatusBadRequest, apierr.CodeInvalidInput, "Invalid message id.")
			return
		}
		var req commodels.PatchMailboxRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			apierr.WriteJSON(w, http.StatusBadRequest, apierr.CodeInvalidInput, "Invalid JSON body.")
			return
		}
		if req.Folder != nil && !validatePatchFolder(*req.Folder) {
			apierr.WriteJSON(w, http.StatusBadRequest, apierr.CodeInvalidInput, "Invalid folder.")
			return
		}
		ok2, err := communication.UpdateMailbox(r.Context(), d.Pool, userID, mid, &req)
		if err != nil {
			apierr.WriteJSON(w, http.StatusInternalServerError, apierr.CodeInternal, "Could not update message.")
			return
		}
		if !ok2 {
			apierr.WriteJSON(w, http.StatusNotFound, apierr.CodeNotFound, "Not found.")
			return
		}
		d.notifyMailbox(userID)
		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(w).Encode(map[string]any{"ok": true})
	}
}

// handleCommUnread is GET /api/v1/communication/unread-count
func (d Deps) handleCommUnread() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet {
			w.Header().Set("Allow", http.MethodGet)
			http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
			return
		}
		userID, ok := d.meUserID(w, r)
		if !ok {
			return
		}
		n, err := communication.CountUnreadInbox(r.Context(), d.Pool, userID)
		if err != nil {
			apierr.WriteJSON(w, http.StatusInternalServerError, apierr.CodeInternal, "Could not count unread.")
			return
		}
		w.Header().Set("Content-Type", "application/json; charset=utf-8")
		_ = json.NewEncoder(w).Encode(commodels.UnreadCountResponse{UnreadInbox: n})
	}
}

// handleCommWS is GET /api/v1/communication/ws — first text message: {"authToken":"…"} (serde camelCase).
func (d Deps) handleCommWS() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet {
			w.Header().Set("Allow", http.MethodGet)
			http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
			return
		}
		if d.JWTSigner == nil {
			http.Error(w, "auth not configured", http.StatusServiceUnavailable)
			return
		}
		c, err := websocket.Accept(w, r, &websocket.AcceptOptions{
			OriginPatterns: []string{"*"},
		})
		if err != nil {
			return
		}
		defer func() { _ = c.Close(websocket.StatusNormalClosure, "") }()

		readAuthCtx, cancelAuth := context.WithTimeout(r.Context(), 2*time.Minute)
		defer cancelAuth()

		typ, b, err := c.Read(readAuthCtx)
		if err != nil {
			return
		}
		if typ != websocket.MessageText {
			return
		}
		var m struct {
			AuthToken string `json:"authToken"`
		}
		if err := json.Unmarshal(b, &m); err != nil || m.AuthToken == "" {
			return
		}
		u, err := d.JWTSigner.Verify(m.AuthToken)
		if err != nil {
			return
		}
		uid, err := uuid.Parse(u.UserID)
		if err != nil {
			return
		}

		runCtx, stop := context.WithCancel(r.Context())
		defer stop()

		if d.Comm != nil {
			recv, unsub := d.Comm.Subscribe()
			defer unsub()
			go func() {
				for {
					select {
					case ev, ok := <-recv:
						if !ok {
							return
						}
						if ev.UserID == uid {
							_ = c.Write(runCtx, websocket.MessageText, []byte(ev.JSON)) //nolint:errcheck
						}
					case <-runCtx.Done():
						return
					}
				}
			}()
		}

		for {
			_, _, err := c.Read(runCtx)
			if err != nil {
				return
			}
		}
	}
}
